 ## 1. Imports e carregamento

In [1]:
import pandas as pd

pd.set_option("display.max_columns", None)

orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")

## 2. Visão geral de cada tabela

In [3]:
for nome, df in [("orders", orders), ("items", items), ("customers", customers),
                  ("products", products), ("payments", payments), ("reviews", reviews)]:
    print(f"\n{'='*40}\n{nome} — shape: {df.shape}\n{'='*40}")
    display(df.head(3))


orders — shape: (99441, 8)


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00



items — shape: (112650, 7)


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87



customers — shape: (99441, 5)


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP



products — shape: (32951, 9)


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0



payments — shape: (103886, 5)


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71



reviews — shape: (99224, 7)


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24


## Panorama Geral

| Tabela   | Linhas | Colunas |
|-----------|-------:|---------:|
| `orders` | 99.441 | 8 |
| `items` | 112.650 | 7 |
| `customers` | 99.441 | 5 |
| `products` | 32.951 | 9 |
| `payments` | 103.886 | 5 |
| `reviews` | 99.224 | 7 |

## 3. Tipos e Nulos

In [4]:
orders.info()
orders.isna().sum()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 21.9 MB


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

## Valores Nulos em `orders`

| Coluna | Valores nulos |
|--------|--------------:|
| `order_approved_at` | 160 |
| `order_delivered_carrier_date` | 1.783 |
| `order_delivered_customer_date` | 2.965 |

## 3. Status dos pedidos

In [5]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

## Distribuição de `order_status`

| Status | Quantidade | Percentual |
|--------|-----------:|------------:|
| `delivered` | 96.478 | 97% |
| `shipped` | 1.107 | 1,11% |
| `canceled` | 625 | 0,63% |
| `unavailable` | 609 | 0,61% |
| `invoiced` | 314 | 0,32% |
| `processing` | 301 | 0,30% |
| `created` | 5 | < 0,01% |
| `approved` | 2 | < 0,01% |


**Insights:** 97% dos pedidos estão delivered e os outros 3% (cancelados, indisponíveis, em processamento)são exatamente os que geram os nulos nas datas de entrega.

**Decisão:** filtrar `order_status == 'delivered'` pode ser uma regra de negócio válida pra camada Silver/Gold ou mantém todos os status com tratamento
de nulo.


## 3. Datas nulas (Pedidos ainda não entregues)

In [6]:
orders[["order_delivered_customer_date", "order_estimated_delivery_date"]].isna().sum()

order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

**Insight:** `order_estimated_delivery_date` (data estimada de entrega) nunca é nula.A Olist sempre calcula uma estimativa no momento da compra, independente do que acontece depois. 

Já `order_delivered_customer_date` (data real de entrega) tem 2.965 nulos. Exatamente o número de pedidos que nunca chegaram a ser entregues ao cliente.

Isso bate certinho com o `order_status`

## Pedidos sem `order_delivered_customer_date`

| Status | Quantidade | Observação |
|--------|-----------:|------------|
| `shipped` | 1.107 | Pedido em trânsito, ainda sem data de entrega ao cliente. |
| `canceled` | 625 | Pedido cancelado antes da entrega. |
| `unavailable` | 609 | Produto indisponível, pedido não concluído. |
| `processing` | 301 | Pedido ainda em processamento. |
| `created` | 5 | Pedido criado, mas ainda não processado. |
| `approved` | 2 | Pagamento aprovado, aguardando próximas etapas. |


Somando `canceled + unavailable + processing + created + approved + shipped = 2.649`, próximo dos 2.965 nulos (a diferença pode ser por pedidos `invoiced` também sem entrega, ou pequenas inconsistências no dado).

**Decisão:**

Isso confirma que o nulo em `order_delivered_customer_date` não é erro de qualidade de dado, é um estado legítimo do pedido ("ainda não chegou"). Duas escolhas de design pra Silver/Gold, ambas válidas (dependendo) do objetivo da análise:

1. Filtrar `order_status == 'delivered'` antes de calcular métricas de tempo de entrega (evita viés, a gente só mede o que de fato foi entregue).

2. Manter todos os status, mas tratar a coluna de `entrega como nula `intencional (útil caso queira analisar taxa de cancelamento/perda, não só tempo de entrega).

Se usarmos `order_estimated_delivery_date` (que nunca é nula) pra outras análises futuras, ela é uma boa candidata pra comparar estimado vs. real. Isso pode virar uma métrica de pontualidade da entrega, um KPIs importante em e-commerce.

## 4. Tempo de entrega

In [7]:
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])
orders["order_delivered_customer_date"] = pd.to_datetime(orders["order_delivered_customer_date"])

orders["tempo_entrega_dias"] = (
    orders["order_delivered_customer_date"] - orders["order_purchase_timestamp"]
).dt.days

orders["tempo_entrega_dias"].describe()

count    96476.000000
mean        12.094086
std          9.551746
min          0.000000
25%          6.000000
50%         10.000000
75%         15.000000
max        209.000000
Name: tempo_entrega_dias, dtype: float64

**Insight:** o count (96.476) bate exatamente com o número de pedidos delivered (96.478, com diferença mínima por arredondamento/edge case). Confirma que o cálculo já está implicitamente restrito aos pedidos entregues, já que `dt.days` retorna `NaN` quando a data de entrega é nula.

A distribuição é bem concentrada: metade dos pedidos (mediana) chega em até 10 dias, e 75% chegam em até 15 dias. Só que a média (12,1) é puxada pra cima da mediana (10),  sinal de assimetria à direita, ou seja, existem pedidos que demoram muito mais que o normal e "puxam" a média.

O `max` de 209 dias é o outlier mais evidente, quase 7 meses pra entregar um pedido é um caso muito fora do padrão (desvio padrão de ~9,5 dias, então 209 está a mais de 20 desvios da média). Vale investigar se é:

- um erro de digitação/carga no dado original,
- um caso real de extravio recuperado tarde,
- ou um pedido com problema logístico grave (ex: endereço errado, greve, etc).

**Decisão:** antes de usar `tempo_entrega_dias` como métrica na camada Gold (ex: tempo médio de entrega por região), vale aplicar algum **tratamento de outlier** na Silver, seja removendo casos extremos (ex: > 60 dias) ou apenas sinalizando-os numa coluna separada, sem descartar o dado bruto.

## 5. Pagamentos

In [8]:
payments["payment_type"].value_counts()
payments["payment_value"].describe()

count    103886.000000
mean        154.100380
std         217.494064
min           0.000000
25%          56.790000
50%         100.000000
75%         171.837500
max       13664.080000
Name: payment_value, dtype: float64

**Insight:** distribuição fortemente assimétrica, a mediana (R$ 100) é bem menor que a média (R$ 154), e o desvio padrão (217) é maior que a própria média, o que é um sinal claro de cauda longa (poucos valores muito altos distorcendo a média). O max de R$ 13.664,08 é um caso extremo, muito acima do 75º percentil (R$ 171,84).

O `min = 0.00` também chama atenção, um pagamento de valor zero é estranho (pode ser voucher, cupom, ou dado inconsistente) e merece checagem separada.

Como `payments` tem uma linha por parcela (não por pedido), os valores
podem precisar ser agregados por `order_id` antes de qualquer análise
de ticket médio.

**Decisão:** pra métricas de "ticket médio" na Gold, a mediana é uma medida mais robusta que a média nesse tipo de distribuição. Também vale investigar os pagamentos de R$ 0 e os valores muito altos antes de decidir se entram nas agregações ou são tratados à parte (ex: poderiam representar múltiplas parcelas somadas por pedido, já que a tabela payments tem uma linha por parcela, não por pedido).

Resumo: usar mediana como medida central em vez de média para "ticket médio" na Gold; investigar pagamentos de valor zero.

Resultado shape, os nulos, e o value_counts() do order_status

## 5. Investigação de outliers - tempo de entrega
1. Outlier de tempo de entrega 

In [9]:
outlier_entrega = orders[orders["tempo_entrega_dias"] > 60].sort_values(
    "tempo_entrega_dias", ascending=False
)
outlier_entrega[[
    "order_id", "order_status", "order_purchase_timestamp",
    "order_delivered_customer_date", "order_estimated_delivery_date",
    "tempo_entrega_dias"
]].head(10)

,order_id,order_status,order_purchase_timestamp,order_delivered_customer_date,order_estimated_delivery_date,tempo_entrega_dias
19590,ca07593549f1816d26a572e06dc1eab6,delivered,2017-02-21 23:31:27,2017-09-19 14:36:39,2017-03-22 00:00:00,209.0
55619,1b3190b2dfa9d789e1f14c05b647a14a,delivered,2018-02-23 14:57:35,2018-09-19 23:24:07,2018-03-15 00:00:00,208.0
61610,440d0d17af552815d15a9e41abe49359,delivered,2017-03-07 23:59:51,2017-09-19 15:12:50,2017-04-07 00:00:00,195.0
89130,285ab9426d6982034523a855f55a885e,delivered,2017-03-08 22:47:40,2017-09-19 14:00:04,2017-04-06 00:00:00,194.0
70307,2fb597c2f772eca01b1f5c561bf6cc7b,delivered,2017-03-08 18:09:02,2017-09-19 14:33:17,2017-04-17 00:00:00,194.0
38509,0f4519c5f1c541ddec9f21b3bddd533a,delivered,2017-03-09 13:26:57,2017-09-19 14:38:21,2017-04-11 00:00:00,194.0
11399,47b40429ed8cce3aee9199792275433f,delivered,2018-01-03 09:44:01,2018-07-13 20:51:31,2018-01-19 00:00:00,191.0
81401,2fe324febf907e3ea3f2aa9650869fa5,delivered,2017-03-13 20:17:10,2017-09-19 17:00:07,2017-04-05 00:00:00,189.0
54480,2d7561026d542c8dbd8f0daeadf67a43,delivered,2017-03-15 11:24:27,2017-09-19 14:38:18,2017-04-13 00:00:00,188.0
62286,437222e3fd1b07396f1d9ba8c15fba59,delivered,2017-03-16 11:36:00,2017-09-19 16:28:58,2017-04-28 00:00:00,187.0


In [10]:
outlier_entrega.shape[0]

288

**Insight:** outliers de tempo de entrega

288 pedidos (cerca de 0,3% do total de entregues) levaram mais de 60 dias.

Parece ser um fechamento de sistemas em lote.

Datas de entrega no topo do ranking:

```text
2017-09-19 14:36:39
2017-09-19 15:12:50
2017-09-19 14:00:04
2017-09-19 14:33:17
2017-09-19 14:38:21
2017-09-19 17:00:07
...
2018-09-19 23:24:07
```

Não é caso isolado é um padrão.

Múltiplos pedidos completamente diferentes, comprados em datas distantes (fevereiro, março de 2017), todos aparecem como "entregues" no mesmíssimo dia: 19/09/2017 (e outro grupo em 19/09/2018, exatamente um ano depois).

Não acredito que o pedido  demorou 200 dias no caminhão. Pode ser que a Olist rodou algum job de "fechamento" ou reconciliação de pedidos travados/perdidos numa data específica, marcando todos como entregues de uma vez.

**Decisão:**
Vale checar se existe uma concentração de datas de entrega (`value_counts() em order_delivered_customer_date`) que confirme esse padrão em lote de forma mais ampla.

Excluir ou sinalizar esses registros como "entrega reconciliada em lote" antes de calcular métricas de SLA/tempo de entrega real na camada Gold, eles distorceriam a análise de performance logística.